apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.cs

```diff
@@ -276,5 +278,190 @@ private static string ValidateTableName(string table, string parameterName)
+                        }
+                        else
+                        {
+                            if (c == s_singleQuote || c == s_doubleQuote || c == s_backtickQuote)
+                            {
+                                isInQuoted = true;
+                                startQuote = c;
+                                // Check whether it is a triple-quote.
+                                if (sql.Length > index + 2
+                                    && sql[index + 1] == startQuote
+                                    && sql[index + 2] == startQuote)
+                                {
+                                    isTripleQuoted = true;
+                                    res.Append(c).Append(c);
+                                    index += 2;
+                                }
+                            }
+                            res.Append(c);
+                        }
+                    }
+                }
+                index++;
+            }
+            if (isInQuoted)
+            {
+                throw new SpannerException(
+                    ErrorCode.InvalidArgument, $"SQL statement contains an unclosed literal: {sql}");
+            }
+            if (res.Length > 0 && res[res.Length - 1] == ';')
+            {
+                res.Remove(res.Length - 1, 1);
+            }
+            return res.ToString();
+        }
+
+        /// <summary>
+        /// First removes all comments from the entire sql string, and then removes any statement hints
+        /// at the start of the sql statement. The returned statement should only be used to determine the
+        /// type of statement. The original statement should always be passed on to Cloud Spanner.
+        /// </summary>
+        /// <param name="sql">The sql statement to strip for comments and statement hints</param>
+        /// <returns>The sql statement without comments and statement hints</returns>
+        internal static string RemoveCommentsAndStatementHint(string sql)
+        {
+            GaxPreconditions.CheckNotNull(sql, nameof(sql));
+            // First remove all comments from the statement.
+            sql = RemoveComments(sql);
+
+            // Valid statement hints at the beginning of a query statement can only contain a fixed set of
+            // possible values. Although it is possible to add a @{FORCE_INDEX=...} as a statement hint, the
+            // only allowed value is _BASE_TABLE. This means that we can safely assume that the statement
+            // hint will not contain any special characters, for example a closing curly brace or one of the
+            // keywords SELECT, UPDATE, DELETE, WITH, and that we can keep the check simple by just
+            // searching for the first occurrence of a keyword that should be preceded by a closing curly
+            // brace at the end of the statement hint.
+            int startStatementHintIndex = sql.IndexOf('{');
```

Comment:
What if the curly bracket is in a string literal?

Reply:
You're right, this can be done a lot simpler, especially when we assume that the statement is valid and well-formed.